arwen/host/execution.go

```diff
@@ -144,6 +144,85 @@ func (host *vmHost) doRunSmartContractCall(input *vmcommon.ContractCallInput) (v
 	return vmOutput
 }
 
+func (host *vmHost) doRunSmartContractUpgrade(input *vmcommon.ContractCallInput) (vmOutput *vmcommon.VMOutput) {
+	host.ClearStateStack()
+	host.InitState()
+
+	runtime := host.Runtime()
+	metering := host.Metering()
+	output := host.Output()
+	storage := host.Storage()
+
+	var err error
+	defer func() {
+		if err != nil {
+			var message string
+			if err == arwen.ErrSignalError {
+				message = output.ReturnMessage()
+			} else {
+				message = err.Error()
+			}
+			vmOutput = output.CreateVMOutputInCaseOfError(output.ReturnCode(), message)
+		}
+	}()
+
+	address := input.RecipientAddr
+
+	contractCode, err := input.GetCodeUpgrade()
+	if err != nil {
+		output.SetReturnCode(vmcommon.UpgradeFailed)
+		return vmOutput
+	}
+
+	contractCodeMetadata, err := input.GetCodeMetadataUpgrade()
+	if err != nil {
+		output.SetReturnCode(vmcommon.UpgradeFailed)
```

Comment:
If you choose to refactor this into a function `performCodeDeploy()`: it will return a `vmOutput` which, in case of failure, contain will contain the return code `vmcommon.ExecutionFailed`. But `doRunSmartContractUpgrade()` could alter the return code from `ExecutionFailed` to `UpgradeFailed`, for this specific case.

Reply:
Refactored, fixed.